src/main/java/com/tabnine/capabilities/CapabilitiesService.java

```diff
@@ -28,27 +28,71 @@ public void init() {
   }
 
   public boolean isCapabilityEnabled(Capability capability) {
-    if (fetchCapabilitiesFuture == null || !fetchCapabilitiesFuture.isDone()) {
-      return false;
+    synchronized (enabledCapabilities) {
+      return enabledCapabilities.contains(capability);
     }
-    return enabledCapabilities.contains(capability);
   }
 
-  private void scheduleFetchCapabilitiesTask() {
-    fetchCapabilitiesFuture =
-        AppExecutorUtil.getAppScheduledExecutorService()
-            .schedule(this::fetchCapabilities, 2, TimeUnit.SECONDS);
+  private synchronized void scheduleFetchCapabilitiesTask() {
+    if (refreshLoop == null) {
+      refreshLoop = new Thread(() -> this.fetchCapabilitiesLoop());
+      refreshLoop.setDaemon(true);
+      refreshLoop.start();
+    }
+  }
+
+  public static final int INITIAL_DELAY_MS = 2000;
+  public static final int LOOP_INTERVAL_MS = 1000;
+
+  public static final int REFRESH_EVERY_MS = 10 * 60 * 1000; // 10 minutes
+
+  private void fetchCapabilitiesLoop() {
+    Optional<Long> lastRefresh = Optional.empty();
+    Optional<Long> lastPid = Optional.empty();
+
+    try {
+      Thread.sleep(INITIAL_DELAY_MS);
+
+      while (true) {
+        try {
+          Long pid = binaryRequestFacade.pid();
+          if (!lastRefresh.isPresent()
+              || System.currentTimeMillis() - lastRefresh.get() >= REFRESH_EVERY_MS
+              || !lastPid.isPresent()
+              || lastPid.get() == null
+              || !lastPid.get().equals(pid)) {
+            fetchCapabilities();
+
+            lastRefresh = Optional.of(System.currentTimeMillis());
+            lastPid = Optional.of(pid);
+          }
+        } catch (Throwable t) {
+          Logger.getInstance(getClass()).warn("Unexpected error. Capabilities refresh failed", t);
+        }
+
+        Thread.sleep(LOOP_INTERVAL_MS);
+      }
+    } catch (Throwable t) {
+      Logger.getInstance(getClass()).warn("Unexpected error. Capabilities refresh loop exiting", t);
+    }
   }
 
   private void fetchCapabilities() {
     final CapabilitiesResponse capabilitiesResponse =
         binaryRequestFacade.executeRequest(new CapabilitiesRequest());
     if (capabilitiesResponse != null && capabilitiesResponse.getEnabledFeatures() != null) {
+      setCapabilities(capabilitiesResponse);
+    } else {
+      scheduleFetchCapabilitiesTask();
```

Comment:
`scheduleFetchCapabilitiesTask` calls `fetchCapabilities`, which then calls `scheduleFetchCapabilitiesTask` in case of failure - why is this kind of recursion needed? 
once `init` is called, its calling `scheduleFetchCapabilitiesTask` which runs in an infinite loop - so why need to re-schedule in case of failure here?

Reply:
in addition:
since you're calling `scheduleFetchCapabilitiesTask` from within the thread, which can, in its turn, spawn another one recursively - this function should also be synchronized on `refreshLoop`
(since it can be called from many threads)